### PR TITLE
contrib: add gorm.io/gorm

### LINKED ISSUE #684: contrib/jinzhu/gorm: support v2
The existing GORM module I am using, https://github.com/jinzhu/gorm, will be V2'd under https://github.com/go-gorm/gorm

I am testing a DataDog-enhanced repo for upgradeability to GORM V2, and would like to know if an adapter for the new GORM is in the works? Looking forward to it :) Thanks!

### PR DESCRIPTION
**Pull request by AdallomRoy**
Fixes #684

### CODE REVIEW

File: contrib/gorm.io/example_test.go
Comment (gbbr) [inline contrib/gorm.io/example_test.go:1]: Please read https://github.com/DataDog/dd-trace-go/tree/v1/contrib#usage to better understand how to name files and folder structure.

File: contrib/gorm.io/gorm.go
Comment (gbbr) [inline contrib/gorm.io/gorm.go:7]: Please use the same package name as the original one.

**Review by gbbr — COMMENTED** [review]
I recommend taking a look at other integrations to figure out what the style conventions and recommendations are for naming etc.

Once it's ready for review, just take it out of draft.

**Comment by AdallomRoy** [comment]
Thank you for your comments. I will first make sure all tests pass and then re-order everything to be according to your coding conventions

**Comment by AdallomRoy** [comment]
I'm pretty sure this is in accordance to your guidelines.
The tests that fail seem unrelated to my work, I'm not sure how stable the tests are normally, but please let me know if you think otherwise.

File: contrib/gorm.io/gorm/gorm.go
Comment (gbbr) [inline contrib/gorm.io/gorm/gorm.go:30]: The idea of integrations is that they use the exact same function signatures as the package being integrated with. In this case, a user can just add a new package without changing any of the existing code and have Datadog running with little to no extra code changes.

The current signature to `Open` is the following:
```
func Open(dialector Dialector, config *Config) (db *DB, err error)
```

File: contrib/gorm.io/gorm/gorm.go
Comment (gbbr) [inline contrib/gorm.io/gorm/gorm.go:31]: Why does this line need to be here?

File: contrib/gorm.io/gorm/gorm.go
Comment (gbbr) [inline contrib/gorm.io/gorm/gorm.go:115]: Please remove most empty lines. The style convention in this repository is mostly following the Go standard library. You will notice that most of the code is compact and uses very little empty lines (if at all)

File: contrib/gorm.io/gorm/gorm.go
Comment (gbbr) [inline contrib/gorm.io/gorm/gorm.go:131]: Can you please go through the PR and remove empty lines?

File: contrib/gorm.io/gorm/gorm.go
Comment (gbbr) [inline contrib/gorm.io/gorm/gorm.go:126]: This one too. The curly braces already create enough spacing for this code not to be cluttered.

File: contrib/gorm.io/gorm/example_test.go
Comment (gbbr) [inline contrib/gorm.io/gorm/example_test.go:20]: We should get rid of this, the example and usage is too complex.

File: contrib/gorm.io/gorm/example_test.go
Comment (gbbr) [inline contrib/gorm.io/gorm/example_test.go:11]: This needs to be aliased as `gormtrace` and the imports need to ordered into 3 groups:
1. Standard library
2. Local (`dd-trace-go`) packages
3. External packages

You will notice this pattern in the other files.

**Review by gbbr — DISMISSED** [review]
Thanks for doing the work. You'll need to bear with us for a few more rounds of reviews but I think we're on the right path.

As for the CI test, no worries, we will look at that separately. It does seem unrelated to your PR. However, you will have to use `gofmt` on your code.

**Comment by gbbr** [comment]
CI is fixed. If you still plan on continuing the PR, once you rebase on `v1` branch it should be fine (for the `ci/circleci: test-contrib` part)

**Comment by AdallomRoy** [comment]
Thanks!

File: contrib/gorm.io/gorm/gorm.go
Comment (AdallomRoy) [reply contrib/gorm.io/gorm/gorm.go:30]: I don't think it's possible to completely mimic the same method signature, since you use sqltrace.Open which needs the driver name, while it's not available in the dialector API. The dialector name can be postgres while the driver name is pgx.

Comment (AdallomRoy) [reply contrib/gorm.io/gorm/gorm.go:30]: What do you suggest to do?

Comment (gbbr) [reply contrib/gorm.io/gorm/gorm.go:32]: What I was suggesting was to move the `sqltraced.Open` part be outside the `Open` function and have the user do that part. Then, the signature could remain the same. Doing that though, I notice that the `Open` function doesn't make much sense anymore so it would likely be better to end up removing it completely and having the user simply call `WithCallbacks` on his own opened database connection (which would probably be better called `Trace` or `RegisterCallbacks`).

I realise this is going now through an experimentation phase to find the best and least intrusive option for a user, but ideally we'd be going through this in the issue before implementing.

What do you think about this?

**Comment by knusbaum** [comment]
Sorry, I've only now had a chance to take a look at this. Previously (in https://github.com/DataDog/dd-trace-go/issues/714) I had looked into making a [`Plugin`](https://pkg.go.dev/gorm.io/gorm#Plugin) to add our tracing, as that looks like a simpler interface for the user. They would simply be able to write:
```
db.Use(gormtrace.Plugin())
```

@AdallomRoy What do you think about a Plugin? Do we lose anything by not using the Dialector interface?

**Comment by AdallomRoy** [comment]
Nice! I can check this out later this week and let you know

**Comment by AdallomRoy** [comment]
re your request, it seems as if plugin initialization happens *after* db.Open, and the DD tracer requires the DB to be opened earlier using the sqltraced.Open, so I think this is not possible with the current design/implementation.
See an example with Prometheus here: https://gorm.io/docs/prometheus.html

Let me know if you think otherwise.

File: contrib/gorm.io/gorm/gorm.go
Comment (AdallomRoy) [reply contrib/gorm.io/gorm/gorm.go:32]: If we ask the user to do the sqltraced.Open himself, then we can also use a plugin like @knusbaum suggested, but then it would require the user to do 2 steps in order to open the DB in a traced manner. I'm not sure which is better.

Comment (gbbr) [reply contrib/gorm.io/gorm/gorm.go:32]: If you don't mind, could you post an example of how that would look like? It doesn't sound too bad based on what you describe, and I think our `jmoiron/sqlx` integration also needs two steps (open, then open again)

**Comment by jumpeiMano** [comment]
Any updates? @AdallomRoy @gbbr

File: contrib/gorm.io/gorm/gorm.go
Comment (AdallomRoy) [reply contrib/gorm.io/gorm/gorm.go:32]: Sorry for the (very) late response.
I think I can make it work and it will look something like this, let me know what you think:
```
// Register augments the provided driver with tracing, enabling it to be loaded by gormtrace.Open.
sqltrace.Register("pgx", &stdlib.Driver{}, sqltrace.WithServiceName("my-service"))
sqlDb, err := sqltrace.Open("pgx", "postgres://pqgotest:password@localhost/pqgotest?sslmode=disable")
if err != nil {
    log.Fatal(err)
}

db, err := gorm.Open(postgres.New(postgres.Config{Conn: sqlDb}), &gorm.Config{})
if err != nil {
    log.Fatal(err)
}

err = db.Use(gormtrace.NewTracerPlugin())
if err != nil {
    log.Fatal(err)
}
```

Do you think this is better than the current API?

**Comment by stlimtat** [comment]
Any updates on this??  I'm keen to use this in my work.

**Comment by AdallomRoy** [comment]
If you are very keen on using it, you can grab it from my repo (using replace in your go.mod), as currently it compiles and all tests pass. - Would be great to get feedback if you are doing so.
Otherwise, you will need to wait until it gets reviewed and merged.

File: contrib/gorm.io/gorm/gorm.go
Comment (knusbaum) [reply contrib/gorm.io/gorm/gorm.go:32]: @AdallomRoy, Thanks for getting back to this.

To me, this new usage is clearer than a function returning a `gorm.Dialector`.

What do you think about also adding a `gormtrace.Open` with the same signature as `gorm.Open` that calls `gorm.Open` followed by `db.Use(...)`. Then, the example usage would be:
```
sqltrace.Register("pgx", &stdlib.Driver{}, sqltrace.WithServiceName("my-service"))
sqlDb, err := sqltrace.Open("pgx", "postgres://pqgotest:password@localhost/pqgotest?sslmode=disable")
if err != nil {
    log.Fatal(err)
}

db, err := gormtrace.Open(postgres.New(postgres.Config{Conn: sqlDb}), &gorm.Config{})
if err != nil {
    log.Fatal(err)
}
```

Comment (AdallomRoy) [reply contrib/gorm.io/gorm/gorm.go]: I updated the API to match your suggestion.
In this case, there is no real need to use gorm Plugins, as this is internal to the tracer anyway, and in my opinion only makes things more complex.
tests & example were fixed to match the new API.

File: contrib/gorm.io/gorm/gorm.go
Comment (knusbaum) [inline contrib/gorm.io/gorm/gorm.go:43]: The documentation for `WithCallbacks` implies that it should be called after `Open`, but `Open` already calls `WithCallbacks`.

We may not even want to export `WithCallbacks`. It is only useful in a case when the user can't change the `gorm.Open` call, in which case they're also probably not able to call `sqltrace.Open` to pass into the dialector config. What do you think? I'm leaning towards making it private.

If we are going to keep `WithCallbacks`, we should:
1. Make the documentation clear that it should only be called on a `*gorm.DB` opened with `gorm.Open`, not `gormtrace.Open`.
2. Add an example for using `WithCallbacks`, since this is not supposed to be used with `gormtrace.Open`.

File: contrib/gorm.io/gorm/gorm.go
Comment (knusbaum) [inline contrib/gorm.io/gorm/gorm.go:61]: Rather than putting `cfg` into the context, I think it is cleaner to just capture it in `afterFunc` and pass it as an argument to `after`. Then we can also avoid the config shuffle in `WithContext`.

**Review by knusbaum — CHANGES_REQUESTED** [review]
@AdallomRoy Thanks, I think this looks a lot nicer usage-wise.
There are a few more changes I think we should make also.

File: contrib/gorm.io/gorm/gorm.go
Comment (knusbaum) [reply contrib/gorm.io/gorm/gorm.go]: I think this looks a lot nicer. I agree, with this method there's no need for a `Plugin`.

The core failure is because `v1.28.0` was released, not something you did. If you upmerge to `v1`, that test should pass.

**Comment by AdallomRoy** [comment]
Pretty sure I'm done with all the CR & formatting requirements. Let me know if there's anything else needed.

File: contrib/gorm.io/gorm/gorm_test.go
Comment (knusbaum) [inline contrib/gorm.io/gorm/gorm_test.go:29]: Please move this up with the previous group (external)

**Review by knusbaum — CHANGES_REQUESTED** [review]
One final import nit.

**Comment by AdallomRoy** [comment]
Thanks - Done

**Review by knusbaum — APPROVED** [review]
LGTM, Thanks @AdallomRoy

**Review by knusbaum — CHANGES_REQUESTED** [review]
@AdallomRoy 
I was doing a final check before merge and realized that the new integration path needs to be versioned according to [the rules here](https://github.com/DataDog/dd-trace-go/blob/v1/contrib/README.md#usage)

It should be just `mv contrib/gorm.io/gorm contrib/gorm.io/gorm.v1`
The package names should remain the same. `example_test.go` import paths will need to change to match.

Sorry for the follow-on task. I'm adding a link in CONTRIBUTING.md to make this clearer. #802

**Comment by AdallomRoy** [comment]
No problem - Done

**Review by knusbaum — APPROVED** [review]
👍  Perfect, thanks!

**Comment by AdallomRoy** [comment]
Short question, this was merged > 1 month ago. Do you have any ETA for 1.29.0? e.g. a stable version containing this feature? Thanks :-)

**Comment by gbbr** [comment]
@AdallomRoy thanks for the ping. I'll bring this up with the team and try to draft a release. Sorry for the long hold

File: contrib/gorm.io/gorm.v1/gorm.go
Comment (hatstand) [inline contrib/gorm.io/gorm.v1/gorm.go:121]: If I understand correctly, this creates a span after an operation has finished. Doesn't this mean that any child spans created during the operation (e.g., for the raw database queries via `database/sql`) will have the wrong parent span?

Comment (gbbr) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: It will be fine, the idea here is that the span has the `tracer.StartTime` option which places the span in the past, to when the operation started.

Comment (hatstand) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: The span would have the correct time yes, but the context isn't updated in `before` so a child span would still have the wrong parent, no?

Comment (gbbr) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: Ah, right. I see what you mean. If someone were to use `db.Statement.Context`. In that case you are right, it wouldn't have _this_ parent. But the question is - would you want it to? Any span started as a child of this span would start after this one finishes, which seems unusual but if you have a valid use case for it we can consider it.

Comment (hatstand) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: Say you've enabled tracing for the `database/sql` connection underlying this `gorm.DB`. Then a child span here (e.g., `postgres.db SELECT * from foo`) would be created that would start and end within the gorm span but not have the gorm span as parent... I think.

Comment (gbbr) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: I'll let the author jump in and hopefully explain the reasoning behind this approach @AdallomRoy

Comment (AdallomRoy) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: Hi @gbbr and @hatstand - Thanks for your comments.
I was trying to create this package as identical as possible to the gorm v1 package here: https://github.com/DataDog/dd-trace-go/blob/v1/contrib/gopkg.in/jinzhu/gorm.v1/gorm.go
so that if you upgrade you will have the same experience.
As far as I understand, this happens in the "after" func in that package as well. Do you think this would behave differently?

Comment (AdallomRoy) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: We could definitely start the span in the before func and save it as part of the context and grab it here for finish if we think that behavior is better

Comment (gbbr) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: My view is that we shouldn't make any premature improvements if this code has had no problems reported so far. As long as there is no issue, let's not fix it 🙂

@hatstand please feel free to open up an issue and describe how to reproduce the problem, and we're happy to look into it. I initially didn't realise that this was just a copy of what was already there.

Comment (hatstand) [reply contrib/gorm.io/gorm.v1/gorm.go:121]: 👍 Reproduced it here https://github.com/DataDog/dd-trace-go/issues/849

**Pull request merged**
